eth_utils/abi.py

```diff
@@ -1,68 +1,781 @@
+        ...       'type': 'string'
+        ...     },
+        ...     {
+        ...       'name': 's',
+        ...       'type': 'uint256'
+        ...     },
+        ...     {
+        ...       'name': 't',
+        ...       'components': [
+        ...         {'name': 'anAddress', 'type': 'address'},
+        ...         {'name': 'anInt', 'type': 'uint256'},
+        ...         {'name': 'someBytes', 'type': 'bytes'},
+        ...       ],
+        ...       'type': 'tuple'
+        ...     }
+        ...   ],
+        ...   'name': 'f',
+        ...   'outputs': [],
+        ...   'payable': False,
+        ...   'stateMutability': 'nonpayable',
+        ...   'type': 'function'
+        ... }
+        >>> get_normalized_abi_inputs(
+        ...   abi, *('myName', 123), **{'t': ('0x1', 1, b'\x01')}
+        ... )
+        ('myName', 123, ('0x1', 1, b'\x01'))
+    """
+    _raise_if_fallback_or_receive_abi(abi_element)
+
+    function_inputs = cast(Sequence[ABIComponent], abi_element.get("inputs", []))
+    if len(args) + len(kwargs) != len(function_inputs):
+        raise TypeError(
+            f"Incorrect argument count. Expected '{len(function_inputs)}'"
+            f", got '{len(args) + len(kwargs)}'."
+        )
+
+    # If no keyword args were given, we don't need to align them
+    if not kwargs:
+        return cast(Tuple[Any, ...], args)
+
+    kwarg_names = set(kwargs.keys())
+    sorted_arg_names = tuple(arg_abi["name"] for arg_abi in function_inputs)
+    args_as_kwargs = dict(zip(sorted_arg_names, args))
+
+    # Check for duplicate args
+    duplicate_args = kwarg_names.intersection(args_as_kwargs.keys())
+    if duplicate_args:
+        raise TypeError(
+            f"{abi_element.get('name')}() got multiple values for argument(s) "
+            f"'{', '.join(duplicate_args)}'."
+        )
+
+    # Check for unknown args
+    # Arg names sorted to raise consistent error messages
+    unknown_args = tuple(sorted(kwarg_names.difference(sorted_arg_names)))
+    if unknown_args:
+        message = "{} got unexpected keyword argument(s) '{}'."
+        if abi_element.get("name"):
+            raise TypeError(
+                message.format(f"{abi_element.get('name')}()", ", ".join(unknown_args))
+            )
+        raise TypeError(
+            message.format(
+                f"Type: '{abi_element.get('type')}'", ", ".join(unknown_args)
+            )
+        )
+
+    # Sort args according to their position in the ABI and unzip them from their
+    # names
+    sorted_args = tuple(
+        zip(
+            *sorted(
+                itertools.chain(kwargs.items(), args_as_kwargs.items()),
+                key=lambda kv: sorted_arg_names.index(kv[0]),
+            )
+        )
+    )
+
+    if len(sorted_args) > 0:
+        return tuple(sorted_args[1])
+    else:
+        return tuple()
+
+
+def get_aligned_abi_inputs(
+    abi_element: ABIElement,
+    args: Union[Tuple[Any, ...], Mapping[Any, Any]],
```

Comment:
Does this need to be updated to be `*args`? How is the Mapping used here, should it be `**kwargs` instead?

Reply:
Looking at these again, I see a couple issues happening with that. I'll get these updated and working, then request another look. Thanks!

Reply:
@fselmo I think we can change the `args` variable to some other name. The mapping (or tuple) consists of the input values given to a function, in the order that is specified by the ABI. A better name might be `input_mappings`, what do you think?

Reply:
Also, `_align_abi_input` should change as well, since it gets called from `get_aligned_abi_inputs`.

Reply:
Digging further from the web3 side, it appears that the `args` provided here are always the result of calls to `get_normalized_abi_inputs`.

Examples:
https://github.com/ethereum/web3.py/pull/3408/files#diff-001703611891fce412dcd3254040ef866df872e607393c7b93ba81aa622f85a8R267
https://github.com/ethereum/web3.py/pull/3408/files#diff-001703611891fce412dcd3254040ef866df872e607393c7b93ba81aa622f85a8R490

Reply:
Use `normalized_args`/`normalized_arg` as arguments for these functions.